src/lib.rs

```diff
@@ -10,30 +10,55 @@ use hyper::http::uri::InvalidUri;
-use lazy_static::lazy_static;
 use std::net::{IpAddr, SocketAddr};
+use std::sync::OnceLock;
 use tokio::io::copy_bidirectional;
 
-lazy_static! {
-    static ref TE_HEADER: HeaderName = HeaderName::from_static("te");
-    static ref CONNECTION_HEADER: HeaderName = HeaderName::from_static("connection");
-    static ref UPGRADE_HEADER: HeaderName = HeaderName::from_static("upgrade");
-    static ref TRAILER_HEADER: HeaderName = HeaderName::from_static("trailer");
-    static ref TRAILERS_HEADER: HeaderName = HeaderName::from_static("trailers");
-    // A list of the headers, using hypers actual HeaderName comparison
-    static ref HOP_HEADERS: [HeaderName; 9] = [
-        CONNECTION_HEADER.clone(),
-        TE_HEADER.clone(),
-        TRAILER_HEADER.clone(),
-        HeaderName::from_static("keep-alive"),
-        HeaderName::from_static("proxy-connection"),
-        HeaderName::from_static("proxy-authenticate"),
-        HeaderName::from_static("proxy-authorization"),
-        HeaderName::from_static("transfer-encoding"),
-        HeaderName::from_static("upgrade"),
-    ];
-
-    static ref X_FORWARDED_FOR: HeaderName = HeaderName::from_static("x-forwarded-for");
+fn te_header() -> &'static HeaderName {
+    static TE_HEADER: OnceLock<HeaderName> = OnceLock::new();
+    TE_HEADER.get_or_init(|| HeaderName::from_static("te"))
+}
+
+fn connection_header() -> &'static HeaderName {
+    static CONNECTION_HEADER: OnceLock<HeaderName> = OnceLock::new();
+    CONNECTION_HEADER.get_or_init(|| HeaderName::from_static("connection"))
+}
+
+fn upgrade_header() -> &'static HeaderName {
+    static UPGRADE_HEADER: OnceLock<HeaderName> = OnceLock::new();
+    UPGRADE_HEADER.get_or_init(|| HeaderName::from_static("upgrade"))
+}
+
+fn trailer_header() -> &'static HeaderName {
+    static TRAILER_HEADER: OnceLock<HeaderName> = OnceLock::new();
+    TRAILER_HEADER.get_or_init(|| HeaderName::from_static("trailer"))
+}
+
+fn trailers_header() -> &'static HeaderName {
+    static TRAILERS_HEADER: OnceLock<HeaderName> = OnceLock::new();
+    TRAILERS_HEADER.get_or_init(|| HeaderName::from_static("trailers"))
+}
+
+fn x_forwarded_for_header() -> &'static HeaderName {
+    static X_FORWARDED_FOR: OnceLock<HeaderName> = OnceLock::new();
+    X_FORWARDED_FOR.get_or_init(|| HeaderName::from_static("x-forwarded-for"))
+}
+
+fn hop_headers() -> &'static [HeaderName; 9] {
+    static HOP_HEADERS: OnceLock<[HeaderName; 9]> = OnceLock::new();
+    HOP_HEADERS.get_or_init(|| {
+        [
+            connection_header().clone(),
+            te_header().clone(),
+            trailer_header().clone(),
+            HeaderName::from_static("keep-alive"),
+            HeaderName::from_static("proxy-connection"),
+            HeaderName::from_static("proxy-authenticate"),
+            HeaderName::from_static("proxy-authorization"),
+            HeaderName::from_static("transfer-encoding"),
+            HeaderName::from_static("upgrade"),
+        ]
+    })
 }
```

Comment:
This got a lot more verbose but I am not sure if there's a better way to do it?

Reply:
Oof, I guess I didn't realize all of this was wrapped up in `lazy_static` too...

Honestly all this stuff might be better off sitting on the `ReverseProxy` struct. That way there's not some kind of Sync call being incurred on every HTTP request (actually, many Sync calls, one per header which is used!).

But I'm happy to merge this like this for now, it's at least not a regression.

Reply:
Hmm, interesting. I don't know a lot about what costs Sync incurs. I guess this is an incentive to get the benchmark file working so we can see if the changes I've made are even observable. I want to learn more about profiling rust code, checking how much memory is allocated for various code paths, etc.